core-bundle/src/Resources/contao/library/Contao/SearchResult.php

```diff
@@ -0,0 +1,130 @@
+<?php
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao;
+
+use Patchwork\Utf8;
+
+/**
+ * Stores result from the Search::query() method
+ *
+ * @author Leo Feyer <https://github.com/leofeyer>
+ */
+class SearchResult
+{
+	/**
+	 * @var array
+	 */
+	private $arrResultsById = array();
+
+	/**
+	 * @var array
+	 */
+	private $arrKeywords;
+
+	/**
+	 * @var array
+	 */
+	private $arrWildcards;
+
+	/**
+	 * @var array
+	 */
+	private $arrPhrases;
+
+	public function __construct(array $arrResults, array $arrKeywords = array(), array $arrWildcards = array(), array $arrPhrases = array())
+	{
+		foreach ($arrResults as $arrRow)
+		{
+			$this->arrResultsById[(int) $arrRow['id']] = $arrRow;
+		}
+
+		$this->arrKeywords = $arrKeywords;
+		$this->arrWildcards = $arrWildcards;
+		$this->arrPhrases = $arrPhrases;
+	}
+
+	public function applyFilter(\Closure $filter): void
+	{
+		foreach ($this->arrResultsById as $intId => $arrRow)
+		{
+			if (!$filter($arrRow))
+			{
+				unset($this->arrResultsById[$intId]);
+			}
+		}
+	}
+
+	public function getCount(): int
+	{
+		return \count($this->arrResultsById);
+	}
+
+	public function getResults(int $intCount = PHP_INT_MAX, int $intOffset = 0): array
+	{
+		$arrResults = \array_slice($this->arrResultsById, $intOffset, $intCount, true);
+		$strIds = implode(',', array_keys($arrResults));
+		$strQuery = 'SELECT * FROM tl_search WHERE ID in (' . $strIds . ')';
+		$objResult = Database::getInstance()->prepare($strQuery)->execute();
+
+		while ($arrRow = $objResult->fetchAssoc())
+		{
+			$arrResults[$arrRow['id']] = array_merge($arrRow, $arrResults[$arrRow['id']]);
```

Comment:
<img width="982" alt="" src="https://user-images.githubusercontent.com/1192057/121702919-92d04e00-cad2-11eb-8cec-40cf17d47112.png">

Is it worth trying to optimize this? Instead of

```php
$ids = [];

while (/* something */) {
    $ids = array_merge($ids, $moreIds);
}
```

we could probably do

```php
$tmp = [];

while (/* something */) {
    $tmp[] = $moreIds;
}

$ids = array_merge(...$moreIds);
```

(I haven't tested the code above.)

Reply:
every single `array_merge()` merges the (full) data from the database with the fields from the existing data (id, protected, groups, and relevance)...

While writing this answer, it would probably be enough to just copy the `relevance` over.

```suggestion
			$arrRow['relevance'] = $arrResults[$arrRow['id']]['relevance'];
			$arrResults[$arrRow['id']] = $arrRow;
```

Reply:
Changed in fd6fa1094a2989d328b2f941406b281402c5876f